redfish-core/include/utils/pcie_util.hpp

```diff
@@ -24,27 +24,90 @@ namespace pcie_util
 {
 
 /**
- * @brief Populate the PCIe Device list from a GetSubTreePaths search of
- *        inventory
+ * @brief Workaround to handle duplicate PCI device list
  *
- * @param[i,o] asyncResp  Async response object
- * @param[i]   Name   Key to store the list of PCIe devices in asyncResp
+ * retrieve PCI device endpoint information and if path is
+ * ~/chassisN/io_moduleN/slotN/adapterN then, replace redfish
+ * PCI device as "chassisN_io_moduleN_slotN_adapterN"
  *
- * @return void
+ * @param[i]   fullPath  object path of PCIe device
+ *
+ * @return string: unique PCIe device name
  */
+inline std::string buildPCIeUniquePath(const std::string& fullPath)
+{
+    sdbusplus::message::object_path path(fullPath);
+    sdbusplus::message::object_path parentPath = path.parent_path();
+    sdbusplus::message::object_path parentofParentPath =
+        parentPath.parent_path();
+    std::string devName;
+
+    if (!parentofParentPath.parent_path().filename().empty())
+    {
+        devName = parentofParentPath.parent_path().filename() + "_";
+    }
+    if (!parentofParentPath.filename().empty())
+    {
+        devName += parentofParentPath.filename() + "_";
+    }
+    if (!parentPath.filename().empty())
+    {
+        devName += parentPath.filename() + "_";
+    }
+    devName += path.filename();
```

Comment:
Is this in a performance sensitive path?

If so, could revert and nest the order of the checks to avoid the parent checks for the levels above the one which is empty. E.g.

if (!parentPath.filename().empty())
{
   if (!parentofParentPath.filename().empty())
  {
     if (!parentofParentPath.parent_path().filename().empty())
     {
        devName = parentofParentPath.parent_path().filename() + "_";
     }
     devName +=...
  }
  devName += ...
}

Reply:
Right. I think we can make the code better.